streams/jvm/src/main/scala/zio/stream/platform.scala

```diff
@@ -302,4 +304,107 @@ trait ZStreamPlatformSpecificConstructors { self: ZStream.type =>
    */
   final def fromJavaStreamTotal[A](stream: => ju.stream.Stream[A]): ZStream[Any, Nothing, A] =
     ZStream.fromJavaIteratorTotal(stream.iterator())
+
+  /**
+   * Create a stream of accepted connection from server socket
+   * Emit socket `Connection` from which you can read / write and ensure it is closed after it is used
+   */
+  def fromSocketServer(
+    port: Int,
+    host: Option[String] = None
+  ): ZStream[Blocking, Throwable, Connection] =
+    for {
+      server <- ZStream.managed(ZManaged.fromAutoCloseable(blocking.effectBlocking {
+                 AsynchronousServerSocketChannel
+                   .open()
+                   .bind(
+                     host.fold(new InetSocketAddress(port))(new InetSocketAddress(_, port))
+                   )
+               }))
+
+      conn <- ZStream.repeatEffect {
+               IO.effectAsync[Throwable, UManaged[Connection]] { callback =>
+                 server.accept(
+                   null,
+                   new CompletionHandler[AsynchronousSocketChannel, Void]() {
+                     self =>
+                     override def completed(socket: AsynchronousSocketChannel, attachment: Void): Unit =
+                       callback(ZIO.succeed(Connection.make(socket)))
+
+                     override def failed(exc: Throwable, attachment: Void): Unit = callback(ZIO.fail(exc))
+                   }
+                 )
+               }
+             } >>= (ZStream.managed(_))
```

Comment:
This is subtly wrong, because it means that as soon as you progress to the next connection, the previous connection is closed. Instead, let's use `ZManaged.scope` here:
```scala
for {
  server <- ZStream.managed(...)
  registerConnection <- ZStream.managed(ZManaged.scope)
  conn <- ZStream.repeatEffect {
                 IO.effectAsync(...).flatMap(managedConn => registerConnection(managedConn).map(_._2))
                }
} yield conn
```

Reply:
Thanks for the tip ! i was not aware of this `ZManaged.scope`